experimental/lib/Transforms/Speculation/HandshakeSpeculation.cpp

```diff
@@ -334,21 +337,34 @@ std::optional<Value> findControlInputToBB(Operation *op) {
   handshake::FuncOp funcOp = op->getParentOfType<handshake::FuncOp>();
   assert(funcOp && "op should have parent function");
 
-  // Find input arcs to BBs in the IR
-  BBtoArcsMap bbToPredecessorArcs = getBBPredecessorArcs(funcOp);
-  unsigned bb = getLogicBB(op).value();
+  std::optional<unsigned> targetBB = getLogicBB(op);
+  if (!targetBB) {
+    op->emitError("Operation does not have a BB.");
+    return {};
+  }
+
+  std::optional<mlir::Value> ctrlSignal = {};
+  for (auto branchOp : funcOp.getOps<handshake::ConditionalBranchOp>()) {
+    if (auto brBB = getLogicBB(branchOp);
+        !brBB || brBB != targetBB)
+      continue;
 
-  // Iterate input arcs to the speculation BB to find the control signal
-  for (const BBArc &arc : bbToPredecessorArcs[bb]) {
-    // Iterate the operands in the edge
-    for (mlir::OpOperand *p : arc.edges) {
-      Value inEdge = p->get();
-      // The control signal should be the only control input to the BB
-      if (inEdge.getType().isa<handshake::ControlType>())
-        return inEdge;
+    if (branchOp.getDataOperand().getType().isa<handshake::ControlType>()) {
+      if (ctrlSignal.has_value()) {
```

Comment:
would it be clearer to have an explicit flag variable rather than this? Feels correct but hard to understand

Reply:
at first I implemented using an explicit variable just before the last commit but I changed it. I can roll it back soon after I return home

Reply:
Oh sorry I'll fix it this weekend